File: core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/messaginglayer/BatchToTupleConverter.scala
package edu.uci.ics.amber.engine.architecture.messaginglayer

import edu.uci.ics.amber.engine.architecture.controller.promisehandlers.WorkerExecutionStartedHandler.WorkerStateUpdated
import edu.uci.ics.amber.engine.architecture.worker.WorkerInternalQueue
import edu.uci.ics.amber.engine.architecture.worker.WorkerInternalQueue.{
  EndMarker,
  EndOfAllMarker,
  InputTuple,
  SenderChangeMarker
}
import edu.uci.ics.amber.engine.common.ambermessage.{
  DataFrame,
  DataPayload,
  EndOfUpstream,
  InputLinking
}
import edu.uci.ics.amber.engine.common.rpc.AsyncRPCClient
import edu.uci.ics.amber.engine.common.statetransition.WorkerStateManager
import edu.uci.ics.amber.engine.common.statetransition.WorkerStateManager.{Ready, Running}
import edu.uci.ics.amber.engine.common.virtualidentity.{
  ActorVirtualIdentity,
  LinkIdentity,
  VirtualIdentity
}

import scala.collection.mutable

class BatchToTupleConverter(
    workerInternalQueue: WorkerInternalQueue,
    workerStateManager: WorkerStateManager,
    asyncRPCClient: AsyncRPCClient
) {

  /**
    * Map from Identifier to input number. Used to convert the Identifier
    * to int when adding sender info to the queue.
    * We also keep track of the upstream actors so that we can emit
    * EndOfAllMarker when all upstream actors complete their job
    */
  private val inputMap = new mutable.HashMap[VirtualIdentity, LinkIdentity]
  private val upstreamMap = new mutable.HashMap[LinkIdentity, mutable.HashSet[VirtualIdentity]]
  private var currentLink: LinkIdentity = _

  private[this] def registerInput(identifier: VirtualIdentity, input: LinkIdentity): Unit = {
    upstreamMap.getOrElseUpdate(input, new mutable.HashSet[VirtualIdentity]()).add(identifier)
    inputMap(identifier) = input
  }

  /** This method handles various data payloads and put different
    * element into the internal queue.
    * data payloads:
    * 1. Data Payload, it will be split into tuples and add to the queue.
    * 2. End Of Upstream, this payload will be received once per upstream actor.
    *    Note that multiple upstream actors can be there for one upstream.
    *    We emit EOU marker when one upstream exhausts. Also, we emit End Of All marker
    *    when ALL upstreams exhausts.
    *
    * @param from
    * @param dataPayload
    */
  def processDataPayload(from: VirtualIdentity, dataPayload: DataPayload): Unit = {
    val link = inputMap(from)
    if (currentLink == null || currentLink != link) {
      workerInternalQueue.appendElement(SenderChangeMarker(link))
      currentLink = link
    }
    dataPayload match {
      case DataFrame(payload) =>
        transitStateToRunningFromReady()
        checkLinkChange(inputMap(from))
        payload.foreach { i =>
          workerInternalQueue.appendElement(InputTuple(i))
        }
      case EndOfUpstream() =>
        transitStateToRunningFromReady()
        val link = inputMap(from)
        checkLinkChange(link)
        upstreamMap(link).remove(from)
        if (upstreamMap(link).isEmpty) {
          workerInternalQueue.appendElement(EndMarker)
          upstreamMap.remove(link)
        }
        if (upstreamMap.isEmpty) {
          workerInternalQueue.appendElement(EndOfAllMarker)
        }
    }
  }

  private def checkLinkChange(link: LinkIdentity): Unit = {
    if (currentLink == null || currentLink != link) {
      workerInternalQueue.appendElement(SenderChangeMarker(link))
      currentLink = link
    }
  }

  private def transitStateToRunningFromReady(): Unit = {
    if (workerStateManager.getCurrentState == Ready) {
      workerStateManager.transitTo(Running)
      asyncRPCClient.send(
        WorkerStateUpdated(workerStateManager.getCurrentState),
        ActorVirtualIdentity.Controller
      )
    }
  }

}

File: core/amber/src/main/python/core/architecture/packaging/batch_to_tuple_converter.py
from collections import defaultdict

from typing import Set, Iterable, Union

from core import Tuple
from core.models.marker import Marker, SenderChangeMarker, EndMarker, EndOfAllMarker
from core.models.payload import DataFrame, EndOfUpstream, DataPayload
from edu.uci.ics.amber.engine.common import ActorVirtualIdentity, LinkIdentity


class BatchToTupleConverter:
    def __init__(self):
        self._input_map: dict[ActorVirtualIdentity, LinkIdentity] = dict()
        self._upstream_map: dict[LinkIdentity, Set[ActorVirtualIdentity]] = defaultdict(set)
        self._current_link: LinkIdentity = None

    def register_input(self, identifier: ActorVirtualIdentity, input_: LinkIdentity) -> None:
        self._upstream_map[input_].add(identifier)
        self._input_map[identifier] = input_

    def process_data_payload(self, from_: ActorVirtualIdentity, data_payload: DataPayload) \
            -> Iterable[Union[Tuple, Marker]]:
        link = self._input_map[from_]
        if self._current_link is None or self._current_link != link:
            self._current_link = link
            yield SenderChangeMarker(link)

        if isinstance(data_payload, DataFrame):
            for tuple_ in data_payload.frame:
                yield tuple_

        elif isinstance(data_payload, EndOfUpstream):
            self._upstream_map[link].remove(from_)
            if len(self._upstream_map[link]) == 0:
                del self._upstream_map[link]
                yield EndMarker()
            if len(self._upstream_map) == 0:
                yield EndOfAllMarker()

        else:
            raise NotImplementedError()